web3/middleware/base.py

```diff
@@ -0,0 +1,129 @@
+        self._w3 = w3
+
+    # -- sync -- #
+
+    def _wrap_make_request(self, make_request: "MakeRequestFn") -> "MakeRequestFn":
+        def middleware(method: "RPCEndpoint", params: Any) -> "RPCResponse":
+            method, params = self.request_processor(method, params)
+            return self.response_processor(method, make_request(method, params))
+
+        return middleware
+
+    def request_processor(self, method: "RPCEndpoint", params: Any) -> Any:
+        return method, params
+
+    def response_processor(self, method: "RPCEndpoint", response: "RPCResponse") -> Any:
+        return response
+
+    # -- async -- #
+
+    async def _async_wrap_make_request(
+        self, make_request: "AsyncMakeRequestFn"
+    ) -> "AsyncMakeRequestFn":
+        async def middleware(method: "RPCEndpoint", params: Any) -> "RPCResponse":
+            method, params = await self.async_request_processor(method, params)
+            return await self.async_response_processor(
+                method,
+                await make_request(method, params),
+            )
+
+        return middleware
+
+    async def async_request_processor(
+        self,
+        method: "RPCEndpoint",
+        params: Any,
+    ) -> Any:
+        return method, params
+
+    async def async_response_processor(
+        self,
+        method: "RPCEndpoint",
+        response: "RPCResponse",
+    ) -> Any:
+        return response
+
+
+class Web3MiddlewareBuilder(Web3Middleware):
+    @staticmethod
+    @abstractmethod
+    def build(
+        w3: Union["AsyncWeb3", "Web3"],
+        *args: Any,
+        **kwargs: Any,
+    ) -> Web3Middleware:
+        """
+        Implementation should initialize the middleware class that implements it,
+        load it with any of the necessary properties that it needs for processing,
+        and curry for the ``w3`` argument since it isn't initially present when building
+        the middleware.
+
+        example implementation:
+
+        ```py
+        class MyMiddleware(Web3BuilderMiddleware):
+            internal_property: str = None
+
+            @staticmethod
+            @curry
+            def builder(user_provided_argument, w3):
+                middleware = MyMiddleware(w3)
+                middleware.internal_property = user_provided_argument
+                return middleware
+
+            def request_processor(self, method, params):
+                ...
+
+            def response_processor(self, method, response):
+                ...
+
+        construct_my_middleware = MyMiddleware.builder
+
+        w3 = Web3(provider)
+        my_middleware = construct_my_middleware("my argument")
+        w3.middleware_onion.inject(my_middleware, layer=0)
+        ```
+        """
+        raise NotImplementedError("Must be implemented by subclasses")
+
+
+# --- type definitions --- #
+
+Middleware = Type[Web3Middleware]
+MiddlewareOnion = NamedElementOnion[str, Middleware]
```

Comment:
All keys (names) in the onion should still be strings, yeah. That logic hasn't changed other than to use the `repr()` of the element if the `TValue` isn't hashable.

Reply:
Actually, I changed that due to [this comment](https://github.com/ethereum/web3.py/blob/45db990ec8240930a11a38e467bfcf200e46aa50/web3/datastructures.py#L32-L39). If `TKey` is bound to a hashable we should define a `__hash__()` that would reflect any changes to the middleware class.

Reply:
False alarm, this was mostly because we borrow the defined ``TKey`` for ``AttributeDict`` data structure. We can revisit this later but we already break these rules in the middleware onion when we set a ``TKey`` as a "name" string. I took out that last commit, adding a newsfragment and merging!